src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceCodeRefactoringProvider.MoveFileCodeAction.cs

```diff
@@ -0,0 +1,162 @@
+                foreach ((var folderName, var index) in candidates)
+                {
+                    if (subFolders.TryGetValue(folderName, out var matchingFolderInfo))
+                    {
+                        var newParts = index >= parts.Length
+                            ? ImmutableArray<string>.Empty
+                            : ImmutableArray.Create(parts, index, parts.Length - index);
+                        var newCurrentFolder = currentFolder.Add(matchingFolderInfo.Name);
+                        builder.AddRange(FindCandidateFolders(matchingFolderInfo, newParts, newCurrentFolder));
+                    }
+                }
+
+                var defaultPathBasedOnCurrentFolder = currentFolder.AddRange(parts);
+                if (builder.All(folders => !folders.SequenceEqual(defaultPathBasedOnCurrentFolder, PathUtilities.Comparer)))
+                {
+                    builder.Add(defaultPathBasedOnCurrentFolder);
+                }
+
+                return builder.ToImmutableAndFree();
+            }
+
+            private class FolderInfo
+            {
+                Dictionary<string, FolderInfo> _folders;
+
+                public string Name { get; }
+
+                public FolderInfo Parent { get; }
+
+                public IReadOnlyDictionary<string, FolderInfo> Folders => _folders;
+
+                private FolderInfo(string name, FolderInfo parent)
+                {
+                    Name = name;
+                    Parent = parent;
+                    _folders = new Dictionary<string, FolderInfo>(StringComparer.Ordinal);
+                }
+
+                public void AddFolder(ImmutableArray<string> folder)
+                {
+                    if (folder.IsDefaultOrEmpty)
+                    {
+                        return;
+                    }
+
+                    var firstFolder = folder[0];
+                    if (!_folders.TryGetValue(firstFolder, out var firstFolderInfo))
+                    {
+                        firstFolderInfo = new FolderInfo(firstFolder, this);
+                        _folders[firstFolder] = firstFolderInfo;
+                    }
+
+                    firstFolderInfo.AddFolder(ImmutableArray.CreateRange(folder.Skip(1)));
+                }
+
+                // TODO: 
+                // Since we are getting folder data from documents, only non-empty folders 
+                // in the project are discovered. It's possible get complete folder structure
+                // from VS but it requires UI thread to do so. We might want to revisit this
+                // later.
+                public static FolderInfo CreateFolderHierarchyForProject(Project project)
+                {
+                    var rootFolderInfo = new FolderInfo("<ROOT>", parent: null);
+                    foreach (var document in project.Documents)
+                    {
+                        rootFolderInfo.AddFolder(document.Folders.ToImmutableArray());
```

Comment:
a more efficient way would just be to use a Queue and Dequeue the head items as you walk down until the quue is empty.

Reply:
Won't fix. Won't make much of a difference in most scenarios.